### PR TITLE
Linter improvements: PRO-2314 PRO-2325 PRO-2330 PRO-2328 PRO-2359 PRO-2313 PRO-2311 PRO-2378

### PR DESCRIPTION
**Pull request by boutell**
- Adds helpful message about using "git diff HEAD" after running the upgrade command.
- Display lint messages in source code order, even if they are for a mix of issues.
- Lint for array field schema properties that need conversion.
- Lint for joinByArray field sub-schema properties that need conversion.
- Lint for joinByOne, joinByArray, joinbyOneReverse, and joinByArrayReverse.
- Lint for widget output method overrides.
- Lint for `filterOptionsForDataAttribute`.
- Lint for methods that should move to the "methods" section.
- Lint for tasks that should move to the "tasks" section.

### CODE REVIEW

File: lib/linter.js [hunk 1]
```diff
@@ -105,20 +107,39 @@ module.exports = ({ argv }) => {
...
-        console.error(' '.repeat(line.columnNumber - 1) + '^');
-        console.error(getMessage({
-          src,
-          line,
-          match
-        }, options.message) + '\n');
-        if (options.url) {
-          console.error(`See: ${options.url}\n`);
-        }
+        results.push({
```
Comment (boutell) [inline lib/linter.js:110]: Build an array of these so I can sort by line number for readability.

File: lib/linters.js [hunk 2]
```diff
@@ -121,9 +121,140 @@ const js = {
...
+      for import purposes (CSV, etc) and may also accept a different representation
+      for form submissions where appropriate. "convert" may be async and now receives
+      these arguments, typically copying from "data[field.name]" to
+      "destination[field.name]" after sanitization:
+
+      "req, field, data, destination"
+    `
+  },
+  joinByArraySchema: {
+    match: /(joinByArray|relationship)[\s\S]{1,1000}schema["']?:/,
```
Comment (boutell) [inline lib/linters.js:241]: Compromise technique to find schema "pretty close" to the field type, as we're not doing strict ecmascript parsing here.

Comment (boutell) [reply lib/linters.js:241]: This worked well in some modules and projects tested.

File: lib/linter.js [hunk 3]
```diff
@@ -105,20 +107,39 @@ module.exports = ({ argv }) => {
...
+          lineNumber: line.lineNumber,
+          // stripIndent just doesn't work here and I don't know why
```
Comment (boutell) [inline lib/linter.js:112]: stripIndent should work, it just doesn't, so I punted these lines hard left. Would like to improve this in future if we can. Still think it's more readable than a whole mess of consecutive console calls, which it replaced.

File: lib/linters.js [hunk 4]
```diff
@@ -121,9 +121,140 @@ const js = {
...
+      new "enableBrowserData" method and implement or extend the "getBrowserData"
+      method to make data available at page load time in the browser, or
+      add data attributes to your markup, or use the REST API after page load time.
+    `
+  },
+  tagsField: {
+    match: /type:\s*["']tags[".]/g,
+    message: stripIndent`
+      There is no "tags" field type in A3. If you need tags for a particular
+      piece type, add a second piece type for that purpose and create a "relationship"
```
Comment (abea) [inline lib/linters.js]: "second" could take an extra cycle to work through. "I already have four piece types. What's the second one?" Simply striking that word would be great.

File: lib/linters.js [hunk 5]
```diff
@@ -121,9 +121,140 @@ const js = {
...
+      field to connect them.
+    `
+  },
+  tagsModule: {
+    match: /apos\.tags|apostrophe-tags/g,
+    message: stripIndent`
+      There is no "apostrophe-tags" module in A3. If you need tags for a particular
+      piece type, add a second piece type for that purpose and create a "relationship"
```
Comment (abea) [inline lib/linters.js]: Same thing here about "second."

**Review by abea — CHANGES_REQUESTED** [review]

File: lib/linters.js (hunk 4 above)
Comment (boutell) [reply lib/linters.js]: Addressed this.

**Review by abea — APPROVED** [review]

**Pull request merged**